File: src/main/java/org/panteleyev/money/app/cells/TransactionSumCell.java
/*
 Copyright (c) 2017-2022, Petr Panteleyev

 This program is free software: you can redistribute it and/or modify it under the
 terms of the GNU General Public License as published by the Free Software
 Foundation, either version 3 of the License, or (at your option) any later version.

 This program is distributed in the hope that it will be useful, but WITHOUT ANY
 WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.

 You should have received a copy of the GNU General Public License along with this
 program. If not, see <https://www.gnu.org/licenses/>.
 */
package org.panteleyev.money.app.cells;

import javafx.geometry.Pos;
import javafx.scene.control.TableCell;
import org.panteleyev.money.app.Styles;
import org.panteleyev.money.model.CategoryType;
import org.panteleyev.money.model.Transaction;
import java.math.RoundingMode;

public class TransactionSumCell extends TableCell<Transaction, Transaction> {
    @Override
    public void updateItem(Transaction transaction, boolean empty) {
        super.updateItem(transaction, empty);

        setAlignment(Pos.CENTER_RIGHT);
        getStyleClass().removeAll(Styles.CREDIT, Styles.DEBIT, Styles.TRANSFER);

        if (empty || transaction == null) {
            setText("");
        } else {
            if (transaction.accountCreditedType() == CategoryType.EXPENSES) {
                getStyleClass().add(Styles.DEBIT);
            } else if (transaction.accountCreditedType() == transaction.accountDebitedType()) {
                getStyleClass().add(Styles.TRANSFER);
            } else {
                getStyleClass().add(Styles.CREDIT);
            }

            setText(transaction.amount().setScale(2, RoundingMode.HALF_UP).toString());
        }
    }
}

File: src/main/kotlin/org/panteleyev/money/cells/TransactionSumCell.kt
/*
 * Copyright (c) 2017, Petr Panteleyev <petr@panteleyev.org>
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * 1. Redistributions of source code must retain the above copyright notice, this
 *    list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright notice,
 *    this list of conditions and the following disclaimer in the documentation
 *    and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */
package org.panteleyev.money.cells

import javafx.geometry.Pos
import javafx.scene.control.TableCell
import org.panteleyev.money.Styles.BLACK_TEXT
import org.panteleyev.money.Styles.BLUE_TEXT
import org.panteleyev.money.Styles.RED_TEXT
import org.panteleyev.money.persistence.CategoryType
import org.panteleyev.money.persistence.SplitTransaction
import org.panteleyev.money.persistence.Transaction
import java.math.BigDecimal

class TransactionSumCell : TableCell<Transaction, Transaction>() {
    override fun updateItem(item: Transaction?, empty: Boolean) {
        super.updateItem(item, empty)
        this.alignment = Pos.CENTER_RIGHT
        if (empty || item == null) {
            text = ""
        } else {
            val amount = item.signedAmount

            styleClass.removeAll(RED_TEXT, BLUE_TEXT, BLACK_TEXT)

            if (item is SplitTransaction) {
                styleClass.add(if (amount.signum() < 0) RED_TEXT else BLACK_TEXT)
            } else {
                var s = BLACK_TEXT
                if (item.accountCreditedType != item.accountDebitedType) {
                    if (item.accountDebitedType == CategoryType.INCOMES) {
                        s = BLUE_TEXT
                    } else {
                        s = RED_TEXT
                    }
                }
                styleClass.add(s)
            }

            val format = if (item is SplitTransaction || item.groupId == 0)
                "%s"
            else
                "(%s)"
            text = String.format(format, amount.setScale(2, BigDecimal.ROUND_HALF_UP).toString())
        }
    }
}